### ISSUE TITLE
Which norm is used by the fgm?

### ISSUE THREAD

**Issue by Apromixately** [tensorflow/cleverhans#701]
Looking at the implementation of fgm(..) in attacks_tf.py I found this:
```
    if ord == np.inf:
        normalized_grad = tf.sign(grad)
        normalized_grad = tf.stop_gradient(normalized_grad)
```
If I interpret this correctly this takes the sign of the gradient along each dimension while the description a few lines down
```
    else:
        raise NotImplementedError("Only L-inf, L1 and L2 norms are "
                                  "currently implemented.")
```
and the condition of the if suggest that the infinity norm should be used for normalizing.

**Comment by goodfeli** [comment]
`fgm` uses the Lp norm with p specified by the `ord` argument. The default value of the `ord` argument is `np.inf` but you can also pass `1` or `2`. Other norms are not implemented.

https://github.com/tensorflow/cleverhans/blob/master/cleverhans/attacks_tf.py#L62

**Comment by Apromixately** [comment]
The code that is used is now at
https://github.com/tensorflow/cleverhans/blob/17d9184bb9b81dc1d2a3b03d197b6d50da243ac0/cleverhans/attacks.py#L409
but I still think it's incorrect. 

Using the infinity norm should normalize the gradient with its maximum coordinate, but instead https://github.com/tensorflow/cleverhans/blob/17d9184bb9b81dc1d2a3b03d197b6d50da243ac0/cleverhans/attacks.py#L411
does
`normalized_grad = tf.sign(grad)`

To give an example, since I might be way off here:
I think normalizing the vector [3, 2, 1] with infinity norm should yield [1, 0.66, 0.33] but tf.sign will give [1, 1, 1].

**Comment by goodfeli** [comment]
This step is computing 

argmax_{eta, max_norm(eta) <= 1} dot(eta, grad)

It's not computing grad / max_norm(eta).

It might be better to call the variable `optimal_step` or something rather than `normalized_grad`.

https://arxiv.org/abs/1412.6572

**Comment by goodfeli** [comment]
More generally, for each ord, it's calculating

argmax(eta, norm(eta, ord) <= 1) dot(eta, grad)

For the L2 norm this actually is grad / norm(eta, 2) but for the other norms it's not.

**Comment by Apromixately** [comment]
Where are you getting this from? I was looking at
```
  if ord == np.inf:
    # Take sign of gradient
    normalized_grad = tf.sign(grad)
    # The following line should not change the numerical results.
    # It applies only because `normalized_grad` is the output of
    # a `sign` op, which has zero derivative anyway.
    # It should not be applied for the other norms, where the
    # perturbation has a non-zero derivative.
    normalized_grad = tf.stop_gradient(normalized_grad)
  elif ord == 1:
    red_ind = list(xrange(1, len(x.get_shape())))
    avoid_zero_div = 1e-12
    avoid_nan_norm = tf.maximum(avoid_zero_div,
                                reduce_sum(tf.abs(grad),
                                           reduction_indices=red_ind,
                                           keepdims=True))
    normalized_grad = grad / avoid_nan_norm
  elif ord == 2:
    red_ind = list(xrange(1, len(x.get_shape())))
    avoid_zero_div = 1e-12
    square = tf.maximum(avoid_zero_div,
                        reduce_sum(tf.square(grad),
                                   reduction_indices=red_ind,
                                   keepdims=True))
normalized_grad = grad / tf.sqrt(square)
```
which does compute normalized_grad in the appropriate way for L1 and L2 normalization but for Linf it doesn't. From https://github.com/tensorflow/cleverhans/blob/17d9184bb9b81dc1d2a3b03d197b6d50da243ac0/cleverhans/attacks.py#L411

**Comment by goodfeli** [comment]
"Where are you getting this from? " -> the arxiv link in my previous comment

"which does compute normalized_grad in the appropriate way for L1 and L2 normalization" -> read my previous comment. For L2, solving the constrained optimization problem is the same as dividing by the norm.

For L1, that looks like a bug.
